### PR TITLE
Remove _{fns} from torch, set correct module for linalg and special fns and add __all__ for other submodules

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81126
* #80520

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -966,3 +965,9 @@ def _register_device_module(device_type, module):
     from . import library
     if not TYPE_CHECKING:
         from . import _meta_registrations
+
+final_all = []
+for elem in __all__:
+    if elem[0] != '_':
```
Comment (anjali411) [inline torch/__init__.py:971]: hack to remove everything imported from `torch._C` with an `_`

File: torch/linalg/__init__.py [hunk 2]
```diff
@@ -16,11 +16,15 @@
                          "For a version of this function that does not synchronize, see :func:`{}`.")
 }
 
+def _set_module_and_doc_str(fn, doc_str):
```
Comment (anjali411) [inline torch/linalg/__init__.py:19]: We should probably use a variant of this function instead of `_add_docstr` at most places. It might be nice to make this a common utility function but errr I didn't want to pass a module arg for every function

**Review by lezcano — APPROVED** [review]
Nice! Thank you @anjali411!

File: torch/__init__.py [hunk 3]
```diff
@@ -966,3 +965,9 @@ def _register_device_module(device_type, module):
     from . import library
     if not TYPE_CHECKING:
         from . import _meta_registrations
+
+final_all = []
```
Comment (albanD) [inline torch/__init__.py:969]: Why is this necessary? didn't we update this file to not have any of these already?

File: torch/linalg/__init__.py (hunk 2 above)
Comment (albanD) [reply torch/linalg/__init__.py:19]: Sounds good!

Small nit if you want to avoid having to change all the functions in the file, you can set this new function as `_add_docstr` locally :D

File: torch/special/__init__.py [hunk 4]
```diff
@@ -2,6 +2,11 @@
 from torch._C import _add_docstr, _special  # type: ignore[attr-defined]
 from torch._torch_docs import common_args, multi_dim_common
 
+def _set_module_and_doc_str(fn, doc_str):
```
Comment (albanD) [inline torch/special/__init__.py:5]: Random idea to make this simpler:
```python
from torch._torch_docs import get_add_docstr_fn

_add_docstr = get_add_docstr_fn("torch.special")

```

**Review by albanD — APPROVED** [review]
Very nice change!
Only small nits to make the change smaller, but good as is as well.

**Comment by anjali411** [comment]
`test_normalize_operator_exahustive` (fx tracing test) starts failing for all linalg and special ops because fx traces the underlying `linalg_{fn}` instead of `{fn}`.

Example:

```
import torch
def fn(A):
  return torch.linalg.det(A)

a = torch.fx.symbolic_trace(fn)
print(a.code)
```

Before:
```
def forward(self, A):
    a = A
    linalg_det = torch._C._linalg.linalg_det(a);  a = None
    return linalg_det
```

After:
```
def forward(self, A):
    a = A
    linalg_det = torch.linalg.linalg_det(a);  a = None
    return linalg_det
 ```  

Test failure:
```
    test_normalize_operator_exhaustive_linalg_cholesky_cpu_float32 errored - num_retries_left: 3
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 848, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 848, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 813, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 970, in only_fn
    return fn(slf, *args, **kwargs)
  File "test_fx_experimental.py", line 1554, in test_normalize_operator_exhaustive
    ref_out = traced(*param_values)
  File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 652, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 275, in __call__
    raise e.with_traceback(None)
AttributeError: module 'torch.linalg' has no attribute 'linalg_cholesky'
```

**Comment by albanD** [comment]
Ho, is the `__name__` attribute on these functions wrong?

**Comment by anjali411** [comment]
@lezcano once cholesky and some other ops that exist both in linalg and torch namespace are removed, can we rename the `linalg_fn_name` ops to just `fn_name`? This should fix the `__name__` issue and also seems like a nice cleanup?

Once that's done, we should also be to land this change, I think?

**Comment by lezcano** [comment]
That makes sense to me. Now, those deprecations would still take months, so I think it makes sense to fix as many of the functions as we can now, and then @IvanYashchuk can update his deprecation PRs with the relevant code fixing the remaining ops.
